src/emu/debug/express.cpp

```diff
@@ -413,43 +413,54 @@ void symbol_table::configure_memory(void *param, valid_func valid, read_func rea
 //  add - add a new u64 pointer symbol
 //-------------------------------------------------
 
-void symbol_table::add(const char *name, read_write rw, u64 *ptr)
+void symbol_table::add(std::string &&name, read_write rw, u64 *ptr)
 {
-	m_symlist.erase(name);
-	m_symlist.emplace(name, std::make_unique<integer_symbol_entry>(*this, name, rw, ptr));
+	std::unique_ptr<symbol_entry> symentry = std::make_unique<integer_symbol_entry>(*this, std::move(name), rw, ptr);
+	internal_add(std::move(symentry));
 }
 
 
 //-------------------------------------------------
 //  add - add a new value symbol
 //-------------------------------------------------
 
-void symbol_table::add(const char *name, u64 value)
+void symbol_table::add(std::string &&name, u64 value)
 {
-	m_symlist.erase(name);
-	m_symlist.emplace(name, std::make_unique<integer_symbol_entry>(*this, name, value));
+	std::unique_ptr<symbol_entry> symentry = std::make_unique<integer_symbol_entry>(*this, std::move(name), value);
+	internal_add(std::move(symentry));
 }
 
 
 //-------------------------------------------------
 //  add - add a new register symbol
 //-------------------------------------------------
 
-void symbol_table::add(const char *name, void *ref, getter_func getter, setter_func setter, const std::string &format_string)
+void symbol_table::add(std::string &&name, void *ref, getter_func getter, setter_func setter, const std::string &format_string)
 {
-	m_symlist.erase(name);
-	m_symlist.emplace(name, std::make_unique<integer_symbol_entry>(*this, name, ref, getter, setter, format_string));
+	std::unique_ptr<symbol_entry> symentry = std::make_unique<integer_symbol_entry>(*this, std::move(name), ref, getter, setter, format_string);
+	internal_add(std::move(symentry));
 }
 
 
 //-------------------------------------------------
 //  add - add a new function symbol
 //-------------------------------------------------
 
-void symbol_table::add(const char *name, void *ref, int minparams, int maxparams, execute_func execute)
+void symbol_table::add(std::string &&name, void *ref, int minparams, int maxparams, execute_func execute)
 {
-	m_symlist.erase(name);
-	m_symlist.emplace(name, std::make_unique<function_symbol_entry>(*this, name, ref, minparams, maxparams, execute));
+	std::unique_ptr<symbol_entry> symentry = std::make_unique<function_symbol_entry>(*this, std::move(name), ref, minparams, maxparams, execute);
+	internal_add(std::move(symentry));
+}
+
+
+//-------------------------------------------------
+//  internal_add - add a newly created symbol_entry
+//-------------------------------------------------
+
+void symbol_table::internal_add(std::unique_ptr<symbol_entry> &&symentry)
+{
+	m_symlist.erase(symentry->name());
+	m_symlist.emplace(std::string(symentry->name()), std::move(symentry));
```

Comment:
Why the extra `std::string`?  It will try to coerce the first parameter to `std::string` anyway when it tries to construct the pair.  Also, why even use this two-step process?  You always want to add/replace an entry if one doesn't exist, and `std::unique_ptr` is both default constructable and move assignable, so you save a hash lookup and potentially an allocation if you do it with a single statement:
```
    m_symlist[symentry->name()] = std::move(symentry);
```

Reply:
Make sense, though the first thing that I ask myself when I see your proposed change is "Is the LHS or the RHS of the assignment evaluated first?  If the RHS is evaluated first, then symentry->name() might fail"

I trust your C++ expertise so I am going to assume that your proposal is safe